base/broadcast.jl

```diff
@@ -431,57 +510,178 @@ julia> (1 + im) ./ Nullable{Int}()
 Nullable{Complex{Float64}}()
 ```
 """
-@inline broadcast(f, A, Bs...) = broadcast_c(f, containertype(A, Bs...), A, Bs...)
+broadcast(f, A, Bs...) = broadcast(f,
+                                   Result{combine_types(A, Bs...)}(),
+                                   A, Bs...)
+
+"""
+    broadcast(f, Broadcast.Result{ContainerType}(), As...)
+
+Specify the container-type of the output of a broadcasting operation.
+You can specialize such calls as
+
+    function Broadcast.broadcast(f, ::Broadcast.Result{ContainerType,Void,Void}, As...) where ContainerType
+        ...
+    end
+"""
+function broadcast(f, ::Result{ContainerType,Void,Void}, A, Bs...) where ContainerType
+    ElType = combine_eltypes(f, A, Bs...)
+    broadcast(f,
+              Result{ContainerType,ElType}(combine_indices(A, Bs...)),
+              A, Bs...)
+end
+
+"""
+    broadcast(f, Broadcast.Result{ContainerType,ElType}(indices), As...)
+
+Specify the container-type, element-type, and indices of the output
+of a broadcasting operation. You can specialize such calls as
+
+    function Broadcast.broadcast(f, r::Broadcast.Result{ContainerType,ElType,<:Tuple}, As...) where {ContainerType,ElType}
+        ...
+    end
+
+This variant might be the most convenient specialization for container types
+that don't support [`setindex!`](@ref) and therefore can't use [`broadcast!`](@ref).
+"""
+function broadcast(f, result::Result{ContainerType,ElType,<:Indices}, As...) where {ContainerType,ElType}
+    if !Base._isleaftype(ElType)
+        return broadcast_nonleaf(f, result, As...)
+    end
+    dest = similar(f, result, As...)
+    broadcast!(f, dest, As...)
+end
+
+# default to BitArray for broadcast operations producing Bool, to save 8x space
+# in the common case where this is used for logical array indexing; in
+# performance-critical cases where Array{Bool} is desired, one can always
+# use broadcast! instead.
+function broadcast(f, r::Result{BottomArray{N},Bool}, As...) where N
+    dest = Base.similar(BitArray, indices(r))
+    broadcast!(f, dest, As...)
+end
+
+# When ElType is not concrete, use narrowing. Use the first element of each input to determine
+# the starting output eltype; the _broadcast! method will widen `dest` as needed to
+# accomodate later values.
```

Comment:
Silly question: Is it possible that this ends up in the situation where the element type is too narrow? E.g. For eltypes like `Union{T,Null}` might the final array end up with eltype `T` or `Null`? (This might prevent `push!` on a vector when you have broadcasted it - no problem in the functional case, though).

Reply:
If I understand your question correctly, you can definitely get that, but it's not unique to this PR. (I *think* this PR doesn't alter the element-type behavior whatsoever.) julia 0.6:
```julia
julia> a = Real[1, 2]
2-element Array{Real,1}:
 1
 2

julia> b = map(identity, a)
2-element Array{Int64,1}:
 1
 2

julia> push!(b, 3.2)
ERROR: InexactError()
Stacktrace:
 [1] convert(::Type{Int64}, ::Float64) at ./float.jl:679
 [2] push!(::Array{Int64,1}, ::Float64) at ./array.jl:617

julia> b = a .+ 1
2-element Array{Int64,1}:
 2
 3
```

The reasoning here is that inference is unfortunately likely to give `Any` occasionally for something that doesn't have to be that wide. So it's good to narrow the outputs where possible to keep instability from propagating.

Reply:
Ok thanks for clarifying. 

I'm wondering if we'll need to be "smarter" about union types in the future, now that they are faster.

Reply:
Yes, that's a common behavior already. That could be problematic if you expected to set a value to `null` in the resulting array. I'm not sure what could be done about this, as we can't know what are the possible types without relying on inference, which we want to avoid. And for things like `map(isnull, x)` we definitely don't want an `Array{Union{T, Null}}` result.

See also https://github.com/JuliaLang/julia/pull/23940.

Reply:
The good news is that there's `broadcast!`, so in the end you don't give anything up by forced narrowing.